Mathlib/Combinatorics/Flow/Basic.lean

```diff
@@ -0,0 +1,230 @@
+import Mathlib.Combinatorics.Digraph.Basic
+
+/-!
+# Network Flows
+
+In this file we introduce network flows. A flow network is as a directed graph where each arc of the
+graph has a fixed capacity. A flow from a source vertex `s` to a sink vertex `t` is an assignment of
+"flow units" to each edge, such that for all vertices other than the source and the sink, the
+incoming flow must equal the outgoing flow (_flow conservation_). The value of such an assignment is
+the amount of flow units leaving the source `s` and arriving at the sink `t`. In practice, we are
+usually interested in finding a flow of maximal value for a given network.
+
+## Main definitions
+
+- `Network`: A flow network assigning capacities to all pairs of vertices. Notably, capacities and
+             non-negative and the capacity from a vertex to itself is zero.
+- `PseudoFlow`, `PreFlow`, and `Flow`: An assignment of flow values for a given `Network`, abiding
+                                       increasingly strict constraints.
+
+## Notation
+
+For flows `f f' : N.PseudoFlow`, we define `f ≤ f'` as the _subflow relation_, indicating that for
+all `u v : V`, `f u v ≤ f' u v`.
+
+## References
+
+See [goldberg_network_1989] for a thorough introduction to the topic.
+
+## Tags
+
+flow, network, max-flow, maxflow, cut, min-cut, mincut, graph, subflow
+-/
+
+universe u_V u_R
+
+open BigOperators
+
+section NetworkDefinition
+
+variable (V : Type u_V) (R : Type u_R) [Zero R] [LE R]
+
+/-- A flow network assigning non-negative capacities to all pairs of vertices. -/
+structure Network where
+  /-- The assigned capacities. -/
+  cap : V → V → R
+  nonneg : ∀ u v, 0 ≤ cap u v
+
+/-- A `Network` where the capacities are only non-zero on arcs of `G`. -/
+structure NetworkOn (G : Digraph V) extends Network V R where
+  cap_eq_zero_of_not_adj : ∀ u v, ¬G.Adj u v → cap u v = 0
+
+end NetworkDefinition
+
+variable {V : Type u_V} {R : Type u_R}
+
+namespace Network
+
+section PseudoFlowDefinition
+
+variable [Zero R] [Preorder R]
+
+/-- A `PseudoFlow` is an assignment of non-negative flow values to each arc of the network, such
+that the assigned amount does not exceed the capacity.
+
+Note that this definition, in contrast to what can sometimes be found in the literature, does
+not require that $f(u, v) = -f(v, u)$. This extra requirement eliminates cycles between two
+vertices and makes some theorem statements shorter. However, since larger cycles are still
+permitted, this usually does not help in proofs. To the contrary, defining a flow becomes more
+cumbersome, sometimes significantly so. -/
+@[ext]
+structure PseudoFlow (N : Network V R) where
+  /-- The assignment of flow values to each arc.
+  Do NOT use. Use the coercion to function instead. -/
+  toFun : V → V → R
+  nonneg' : ∀ u v, 0 ≤ toFun u v
+  capacity' : ∀ u v, toFun u v ≤ N.cap u v
+
+namespace PseudoFlow
+
+variable {N : Network V R}
+
+instance instFunLike : FunLike N.PseudoFlow V (V → R) where
```

Comment:
After this was suggested, I split `Flow` into the three structures. Now, I still would want the following lemmas to hold, but coercions fail:
```lean
lemma foo (f : N.PreFlow s t) : True := .intro
example {s t : V} (f : N.Flow s t) := foo f
example {s t : V} (f : N.Flow s t) : R := f s t
```
I fiddled around with `Coe`, `CoeOut` and `CoeFun` a bit, but couldn't get it to work. Any ideas?

Reply:
You need to write one `FunLike` instance per structure